doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,965 @@
+- ``string``, mutable strings
+- ``'a array``, mutable arrays
+- ``exn``, exceptions
+- ``constr``, kernel-side terms
+- ``pattern``, term patterns
+- ``ident``, well-formed identifiers
+
+Type declarations
+~~~~~~~~~~~~~~~~~
+
+One can define new types by the following commands.
+
+.. cmd:: Ltac2 Type @ltac2_typeparams @lident
+   :name: Ltac2 Type
+
+   This command defines an abstract type. It has no use for the end user and
+   is dedicated to types representing data coming from the OCaml world.
+
+.. cmdv:: Ltac2 Type {? rec} @ltac2_typeparams @lident := @ltac2_typedef
+
+   This command defines a type with a manifest. There are four possible
+   kinds of such definitions: alias, variant, record and open variant types.
+
+   .. productionlist:: coq
+      ltac2_typedef    : `ltac2_type`
+                       : [ `ltac2_constructordef` | ... | `ltac2_constructordef` ]
+                       : { `ltac2_fielddef` ; ... ; `ltac2_fielddef` }
+                       : [ .. ]
+      ltac2_constructordef   : `uident` [ ( `ltac2_type` , ... , `ltac2_type` ) ]
+      ltac2_fielddef         : [ mutable ] `ident` : `ltac2_type`
+
+   Aliases are just a name for a given type expression and are transparently
+   unfoldable to it. They cannot be recursive. The non-terminal
+   :production:`uident` designates identifiers starting with an uppercase.
+
+   Variants are sum types defined by constructors and eliminated by
+   pattern-matching. They can be recursive, but the `rec` flag must be
+   explicitly set. Pattern-maching must be exhaustive.
+
+   Records are product types with named fields and eliminated by projection.
+   Likewise they can be recursive if the `rec` flag is set.
+
+   .. cmdv:: Ltac2 Type @ltac2_typeparams @qualid := [ @ltac2_constructordef ]
+
+      Open variants are a special kind of variant types whose constructors are not
+      statically defined, but can instead be extended dynamically. A typical example
+      is the standard `exn` type. Pattern-matching must always include a catch-all
+      clause. They can be extended by this command.
+
+Term Syntax
+~~~~~~~~~~~
+
+The syntax of the functional fragment is very close to the one of Ltac1, except
+that it adds a true pattern-matching feature, as well as a few standard
+constructions from ML.
+
+.. productionlist:: coq
+   ltac2_var        : `lident`
+   ltac2_qualid     : ( `modpath` . )* `lident`
+   ltac2_constructor: `uident`
+   ltac2_term       : `ltac2_qualid`
+                    : `ltac2_constructor` `ltac2_term` ... `ltac2_term`
+                    : `ltac2_term` `ltac2_term` ... `ltac2_term`
+                    : fun `ltac2_var` => `ltac2_term`
+                    : let `ltac2_var` := `ltac2_term` in `ltac2_term`
+                    : let rec `ltac2_var` := `ltac2_term` in `ltac2_term`
+                    : match `ltac2_term` with `ltac2_branch` ... `ltac2_branch` end
+                    : `int`
+                    : `string`
```

Comment:
Why are `int` and `string` terms?

Reply:
Hum, they are just literals I guess. What's surprising?

Reply:
Sorry, I don't get it. Since types are not terms in ML languages, I don't understand in which case this is useful. Furthermore, if `int` and `string` are terms, why not `constr` and `ident`?

Reply:
I agree that they shouldn't be part of the syntax, they're already covered by literals. And this is not PHP (nor Ltac), they don't have any specific status parsingwise.

Reply:
@Zimmi48 I think you are misreading it. These are non-terminals.

@ppedrot I haven't checked if they were redundant, i.e. if literals were already included somewhere else. I haven't added these two, though, AFAIR, so it comes from the previous doc.

Reply:
Ah! That's a misunderstanding actually. They stand for literal integers and strings, respectively. The original doc had probably some markup around.

Reply:
> Ah! That's a misunderstanding actually. They stand for literal integers and strings, respectively. The original doc had probably some markup around.

I'm lost. Isn't still the meaning they have in this doc?

Reply:
I don't know what they're supposed to mean in the Sphinx-Coq grammar description language, but what I meant is that the intended meaning is indeed literals. What the correct way to write this down in the DSL is, I have no idea.

Reply:
Literals and not non-terminals? Because indeed, Maxime's explanation did make sense to me (and corresponds indeed to the fact that there are backquotes around).